cpu/nrf52/radio/nrf802154/nrf802154_radio.c

```diff
@@ -230,11 +230,10 @@ static int _read(ieee802154_dev_t *dev, void *buf, size_t max_size,
             radio_info->lqi = (uint8_t)(hwlqi > UINT8_MAX/ED_RSSISCALE
                                        ? UINT8_MAX
                                        : hwlqi * ED_RSSISCALE);
-            /* Calculate RSSI by subtracting the offset from the datasheet.
-             * Intentionally using a different calculation than the one from
-             * figure 122 of the v1.1 product specification. This appears to
-             * match real world performance better */
-            radio_info->rssi = _hwval_to_ieee802154_dbm(hwlqi) + IEEE802154_RADIO_RSSI_OFFSET;
+            /* We calculate RSSI from LQI, since it's already 8-bit
+               saturated */
```

Comment:
Maybe I'm missing context, but this comment confuses me more than it helps. I also don't quite understand how RSSI can be calculated from LQI. There doesn't seem to be a consistent definition (at least when also considering non-802.15.4 devices) for the LQI. But it seems to be mostly based on bit error rates, on "how easy a received signal can be demodulated" (citing the CC1101 datasheet), or the signal to noise ratio.

So as far as I understood, the LQI can be quite good for a weak signal (if both receiver and transmitter perform optimal and no interference, noise, scattering, reflection or other effects impacting the signal quality are present). Similar, a high RSSI doesn't necessarily mean a high LQI: E.g. when the receiver is right next to the sender, but also right next to an interfering device (e.g. an 802.11 device) the link quality can be poor.

So to my best knowledge, RSSI and LQI are orthogonal to each other and one cannot compute one from the other.

Reply:
@jia200x mind suggesting a better comment since I took this from your suggestion :)

Reply:
this comes from the datasheet:
```
When a packet is received a link quality indicator (LQI) is also generated and appended immediately after the last received octet. When using
IEEE 802.15.4 compliant frame this will be just after the MSDU since the FCS is not reported. In the case of a non-complient frame it will be
appended after the full frame. The LQI reported by hardware must be converted to IEEE 802.15.4 range by an 8-bit saturating multiplication by 4,
as shown in the code example for ED sampling. The LQI is only valid for frames equal to or longer than three octets. When receiving a frame the
RSSI (reported as negative dB) will be measured at three points during the reception. These three values will be sorted and the middle one selected
(median 3) for then to be remapped within the LQI range. The following figure illustrates the LQI measurement and how the data is arranged in
the DataRAM:
```

Reply:
Maybe we can add a reference to the datasheet for this?

Reply:
> So to my best knowledge, RSSI and LQI are orthogonal to each other and one cannot compute one from the other.

This was my understanding too. I agree it looks weird. That's what I interpret from the datasheet at least.

Reply:
Hmm, to me it seems they are using LQI and RSSI as synonyms in the data sheet. IMO it is best to only set the RSSI value and leave the LQI value unset. For other transceivers reporting LQI in addition to RSSI provides additional information, as one cannot be converted to the other. This is here not the case. And this is also consistent with the code example linked in the [data sheet](https://infocenter.nordicsemi.com/pdf/nRF52840_PS_v1.1.pdf) on page 319: It returns only a single value and the function is named `uint8_t sample_ed(void)`. And ED is introduced as abbreviation of Energy Detection just above. So this really seems to be just the signal strength, not the link quality.

If e.g. a routing algorithm in a mesh network has to decide on costs for each link, it might base this on RSSI, LQI, and packet error rate, but ignore RSSI and LQI if those are not available. If the driver reports signal strength via both RSSI and LQI (just in different representation), the cost estimation will be biased towards stronger signals and might result in higher packet error rate.

Reply:
I have the same impression that they are using LQI wrong. I would be OK with not reporting LQI, because I agree a node might receive data with a poor RSSI but a high SNR.

Reply:
But how would a caller know that the lqi is not reported?

Reply:
this one got merged: https://github.com/RIOT-OS/RIOT/pull/15754. I suggest to use that one

Reply:
Note that `lqi` is set to `GNRC_NETIF_HDR_NO_LQI` in `gnrc_netif_hdr_init()` (or via `gnrc_netif_hdr_build()`, which calls in turn `gnrc_netif_hdr_init()`. So if you leave `lqi` untouched, it will have that value already.

Reply:
> Note that `lqi` is set to `GNRC_NETIF_HDR_NO_LQI` in `gnrc_netif_hdr_init()` (or via `gnrc_netif_hdr_build()`, which calls in turn `gnrc_netif_hdr_init()`. So if you leave `lqi` untouched, it will have that value already.

I would rather not use `GNRC_NETIF_HDR_NO_LQI` this is not related to `gnrc`, so this value is only correctly set if gnrc is used.

Reply:
You are right, I missed that :/. I think the cleanest way would be to define a IEEE8021_RADIO_NO_LQI. Then the upper layer can make whatever conversion (e.g `gnrc_netif_ieee802154` can check if there's LQI or not...).

Reply:
alternatively, we can introduce a CAP. That would help to optimize the binary later

Reply:
From the standard I'm not convinced the value is wrong [iee802154-2020 p.457](https://sci-hub.se/downloads/2020-07-27/06/ieee-standard-for-lowrate-wireless-networks.pdf?rand=5fff1f10b3853):

```
10.2.7 Link quality indicator (LQI)
The LQI measurement is a characterization of the strength and/or quality of a received packet. The
measurement may be implemented using receiver ED, a signal-to-noise ratio estimation, or a combination of
these methods. The use of the LQI result by the network or application layers is not specified in this
standard.
The LQI measurement shall be performed for each received packet. The minimum and maximum LQI
values (0x00 and 0xff) should be associated with the lowest and highest quality compliant signals detectable
by the receiver, and LQI values in between should be uniformly distributed between these two limits. At
least eight unique values of LQI shall be used.
```

Reply:
Ok, then let's leave it as it is now. I had the same impression as @maribu , but I htink the standard is the reference here.
I know that comment about LQI and RSSI was written somewhere (at86rf2xx?), but I think it will be better to follow the standard here

Reply:
@maribu are you ok with that? :)

Reply:
I added the RM reference to the comment.